rset.1
.Dd October 8, 2026
.Dt RSET 1
.Os
.Sh NAME
.Nm rset
.Nd remote staging and execution tool
.Sh SYNOPSIS
.Nm rset
.Op Fl AnRt
.Op Fl E Ar environment
.Op Fl F Ar sshconfig_file
.Op Fl f Ar routes_file
.Op Fl x Ar label_pattern
.Ar hostname ...
.Nm rset
.Op Fl E Ar environment
.Op Fl F Ar sshconfig_file
.Op Fl f Ar routes_file
.Op Fl x Ar label_pattern
.Fl o Ar log_directory
.Fl p Ar workers
.Ar hostname ...
.Nm rset
.Ar --
.Ar command
.Sh DESCRIPTION
.Nm
evaluates script fragments written in Progressive Label Notation
.Pq Xr pln 5 ,
executing each script in sequence using ssh connection multiplexing.
.Pp
.Nm
expects one or more hostnames matching labels in the
.Ar routes_file .
If arguments contain characters consistent with a valid network address, they
are compared with all host aliases, otherwise they are treated as regular
expressions.
.Ss Arguments
.Bl -tag -width Ds
.It Fl A
Download files listed in label export paths to the local
.Pa _archive
directory in the format
.Sq hostname:basename(filename) .
Absolute paths are permitted, or paths relative to the staging directory.
.It Fl E Ar environment
Set one or more environment variables using the format
.Sq name="value" ... .
.It Fl F Ar sshconfig_file
Specify a
.Pa config
file for
.Xr ssh 1
to use.
.It Fl f Ar routes_file
Specify a
.Ar routes_file .
This file is composed of labels which correspond to
hostnames or IP addresses which are matched against
.Ar hostname .
The default is
.Pa routes.pln .
.It Fl n
Do not connect to remote hosts.
May be combined with
.Fl x
to highlight label names that match.
Special text defined between { and } is still executed locally.
.It Fl o Ar log_directory
Log directory to use for background workers.
Each log file is created for each worker using the format
.Ql YYYY-MM-DD_HHMMSS.n .
.It Fl p Ar workers
Parallel execution distributed across the specified number of workers.
The log directory must also be specified using
.Fl o .
A summary of results is displayed using the
.Pa rexec-summary
script located in the search path.
.It Fl R
Send files listed in label export paths from the local
.Pa _archive
directory.
This is the reverse of
.Fl A .
.It Fl t
Allow TTY input by copying the content of each label to the remote host instead
of opening a pipe to the interpreter.
.It Fl x Ar label_pattern
Execute labels matching the specified regex.
By default only labels beginning with [0-9a-z] are evaluated.
.El
.Ss Commands
.Bl -tag -width Ds
.It Cm init Op Ar prefix
Initialize a new repository based examples, typically installed in
.Pa /usr/share/examples/rset/ .
Optionally, use an example matching the specified
.Ar prefix ,
.Ql alpine
is the default.
.It Cm wenv
Print shell environment used by parallel workers.
Provides a configuration template for custom status messages.
.El
.Sh ENVIRONMENT
.Ss Status Format
Status messages for each stage of execution may be customized by setting
.Ev RSET_HOST_CONNECT ,
.Ev RSET_HOST_CONNECT_ERROR ,
.Ev RSET_LABEL_EXEC_BEGIN ,
.Ev RSET_LABEL_EXEC_END ,
.Ev RSET_LABEL_EXEC_ERROR ,
and
.Ev RSET_HOST_DISCONNECT .
.Pp
Setting
.Ev RSET_HOST_CONNECT
disables the default status messages and is the only required variable.
These have no effect in conjunction with the
.Fl n
flag.
The following variables are interpolated:
.Pp
.Bl -tag -compact -width 4n
.It Li \%%e
exit code
.It Li \%%l
label name
.It Li \%%h
hostname
.It Li \%%s
session identifier
.It Li \%%T
rfc-3339 timestamp
.It Li \%%%
literal
.Ql %
.El
.Ss Diagnostics
.Bl -tag -width "RSET_ENVIRONMENT"
.It Ev HTTP_TRACE
If defined, print log messages from
.Xr miniquark 1 .
.It Ev SSH_TRACE
If defined, print commands used for remote execution.
.El
.Ss Remote Execution
The following environment variables are set on the remote host:
.Bl -tag -width "RSET_ENVIRONENT"
.It Ev INSTALL_URL
Used by
.Xr rinstall 1
for fetching content from
.Pa _sources .
.It Ev SD
Staging directory where
.Nm
unpacks utilities and configuration files.
.El
.Pp
State may set for the duration of the ssh session using
.Xr renv 1
using
.Ql name=value
as the first argument.
.Sh OPTIONS
The following options are recognized when parsing a
.Xr pln 5
file:
.Bl -tag -width Ds
.It Dv environment=
Space separated list of environment variables validated by
.Xr renv 1 .
.It Dv environment_file=
Space separated list of files defining environment variables validated by
.Xr renv 1 .
.It Dv interpreter=
The interpreter used to run scripts on the remote host.
Default is
.Pa /bin/sh .
.It Dv local_interpreter=
The interpreter to evaluate content that is prepended to a script
Default is
.Pa /bin/sh .
.It Dv execute_with=
Command for elevating privileges, such as
.Xr doas 1
and
.Xr sudo 8 .
.El
.Sh SEE ALSO
.Xr rinstall 1 ,
.Xr rsub 1 ,
.Xr ssh_config 5 ,
.Xr re_format 7
.Sh AUTHORS
.An Eric Radman Aq Mt ericshane@eradman.com


NEWS
= Release History

== 3.5: October 8, 2026

 - Dynamically allocate route labels; use arbitrary limit of 10000 routes
 - Set a limit of 1000 for range expansion
 - New 'init' command with example configurations for 'alpine' and 'openbsd'
 - New 'wenv' command, use custom log format to supersede '-e' flag
 - Display multibyte characters in error messages
 - Enable environment_file= to accept multiple pathnames
 - pln(5): no longer document single-use options
 - pln(5): rewrite man page and organize in three major sections
 - renv(1): require closing quote, reject trailing whitespace

== 3.4: August 3, 2026

 - Match route labels using regular expressions, or any alias using exact hostnames
 - Do not count zero-width regex as a match
 - All utilities: add usage hint and '-h' flag to display option summary
 - Detect string copy overflow and abort with an error
 - rinstall(1) create directories using the same owner as ${SD}
 - miniquark(1) compare HTTP If-Modified-Since times using GMT
 - miniquark(1) adjust options to resemble nc(1)
 - Use 'ustar' format for transfering files on OpenBSD
 - Replace 'tmppath' pledge with 'wpath cpath' on OpenBSD
 - Exit without cleanup if LOCAL_CONTROL_SOCKET already exists
 - Wait 0.5 seconds for SSH control socket to clear

== 3.3: January 17, 2026

 - Use session_id for remote stage dir name
 - Access miniquark logs using HTTP_TRACE
 - Access ssh invocation using SSH_TRACE
 - Do not set INSTALL_URL, rely on default in rinstall(1)
 - Exit on failure for malloc(2), pipe(2), socket(2), bind(2), strdup(2)
 - Allow export path on host labels used for archive/restore
 - pln(5) Ensure export paths are only specified on route labels
 - pln(5) Raise error for indented text in unexpected context
 - miniquark(1) listen on multiple addresses, log each bind address
 - miniquark(1) remove chroot feature
 - rinstall(1) default to curl

== 3.2: February 6, 2025

 - Update MANPREFIX for FreeBSD
 - Format source using clang-format and rubocop
 - rsub(1) allow diff(1) behavior to be modified using RSUB_DIFF_ARGS
 - rsub(1) refactor based on patterns used in rinstall(1)
 - rsub(1) remove temporary source file
 - Ensure top-level directory permissions are not world-readable
 - Avoid leaving a NUL character at the end of local execution script
 - Use native strtonum(3) on MacOS

== 3.1: September 24, 2024

 - Add begin= and end= options
 - Add archive and restore options
 - Allow IPv6 addresses to be used in routes
 - Add out-of-bounds check to options
 - Use 'set -a' instead of 'sh -a' to avoid environment contamination
 - Enable parallel execution using background workers
 - rinstall(1) print name of missing source with absolute path
 - rinstall(1) adjust comments and error messages

== 3.0: February 26, 2024

 - Check for ssh-agent before updating _rutils
 - Correct use of getopt(3) to allow compact flags
 - Use RSET_HOST_CONNECT_ERROR to format failed connection or stagedir initialization
 - Use RSET_LABEL_EXEC_ERROR if ssh(1) or sh(1) terminates during script execution
 - Add log format variable '%s' for generated session ID
 - Remove '-l' option
 - Tests: use curl(1) on Linux
 - Detect and use native strlcpy(3), arc4random(3) on Linux
 - Upload _rutils and exported paths in two steps for compatibility with busybox tar(1)
 - rsub(1): escape ampersands found in '-l' argument
 - rinstall(1) allow diff(1) behavior to be modified using RINSTALL_DIFF_ARGS
 - rinstall(1) add '-a' option for defining an alternate source location
 - rinstall(1) saving a file in $SD no longer depends the current directory
 - rinstall(1) failure running diff(1) no longer reports "files are different"

== 2.9: December 22, 2023

 - Add flag '-E' for overriding remote environment variables
 - rset(1)/renv(1) extended to store session-level state using name="value"
   arguments
 - renv(1) improved error handling, one variable per line
 - Immediately raise errors and abort if environment or environment_file options
   contain syntax errors
 - rinstall(1) downloads sources to the staging directory without a temporary
   file to avoid duplication
 - rinstall(1) prioritize curl with fallback to wget on Linux
 - rinstall(1) recreates file path when fetching over HTTP
 - rinstall(1) allows empty files to be installed
 - Only support '-l' flag with dry-run
 - New custom log format for each stage of execution defined by environment
   variables

== 2.8: November 13, 2023

 - No longer set ROUTE_LABEL, and LABEL on remote host
 - New renv(1) utility for validating environment variables
 - New environment_file option for defining variables
 - New environment option for defining variables inline
 - Improve parsing by prioritizing parameters before labels
 - '-l' requires an option value

== 2.7: October 28, 2023

 - Avoid opening temporary files twice
 - Connect to the hostname specified, not the primary label name
 - Do not start web server during dry run
 - Execute all hostnames specified on the command line
 - Support for hostname sequence ranges in routes file

== 2.6: September 6, 2023

 - rinstall: use wget with fallback to curl except for *BSD
 - configure: use TARGET_OS to override the output of uname(1)
 - MacOS: use tar -c --no-xattrs
 - Use mkstemp(3) to avoid compiler warnings
 - Add bundler/inline and install webrick locally to run tests

== 2.5: December 16, 2022

 - rinstall: ignore environment variable http_proxy
 - rinstall: Use shell getopts to parse arguments
 - Raise error if a label alias begins with a space

== 2.4: May 3, 2022

 - Formatting improvements to tests
 - Route labels may be a comma-separated list

== 2.3: November 17, 2021

 - Tweaks to unveil(2)
 - rinstall(1) accepts a directory instead of a target, similar to cp(1)

== 2.2: June 16, 2021

 - Raise an error if the output of local execution does not end with '\n'
 - Rename "Remote Sequential Execution Tool" to "Remote Staging Execution Tool"
 - Fix miniquark arg parsing on arm

== 2.1: April 3, 2021


rset.c
/*
 * rset.c
 * Configure systems using any scripting language
 */

#include <sys/wait.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <libgen.h>
#include <regex.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "missing/compat.h"

#include "config.h"
#include "execute.h"
#include "input.h"
#include "rutils.h"
#include "worker.h"
#include "xlibc.h"

/* forwards */
static void handle_exit(int sig);
static void usage(bool);
static char **set_options(int argc, char *argv[]);
static void compare_argv(char *args[], char *hostnames[], char *m_args[]);
static void not_found(char *name);
static void start_http_server(int stdout_pipe[], int http_port);
static int execute_remote(char *hostnames[], regex_t *label_reg);
static int dry_run(char *hostnames[], char *m_args[], regex_t *label_reg);
static int init_repository(const char *);

/* globals from input.h */
Label **route_labels;

/* globals */
int archive_opt;
int dryrun_opt;
int restore_opt;
int tty_opt;
int n_parallel;
char *sshconfig_file;
char *env_override;
char *log_directory;
char *label_pattern = DEFAULT_LABEL_PATTERN;
char *routes_file = ROUTES_FILE;

enum {
	cmd_wenv = 1,
	cmd_init = 2,
} command_opt;

/* globals used by signal handlers */
char *socket_path;
char *hostname;
int http_port;

/*
 * Remote Staging Execution Tool
 * configure systems using any scripting language
 */
int
main(int argc, char *argv[]) {
	int fd;
	int i, j;
	int ret;
	int n_workers;
	int n_routes;
	int worker_argc[MAX_WORKERS];
	int worker_pid[MAX_WORKERS];
	char *renv_bin, *rinstall_bin, *rsub_bin;
	char **args, **hostnames, **m_args;
	char **worker_argv[MAX_WORKERS];
	char routes_realpath[PATH_MAX];
	regex_t label_reg;
	struct sigaction act;

	/* terminate SSH connection if a signal is caught */
	act.sa_flags = 0;
	act.sa_flags = SA_RESETHAND;
	act.sa_handler = handle_exit;
	if (sigemptyset(&act.sa_mask) & (sigaction(SIGINT, &act, NULL) != 0))
		err(1, "Failed to set SIGINT handler");
	if (sigemptyset(&act.sa_mask) & (sigaction(SIGTERM, &act, NULL) != 0))
		err(1, "Failed to set SIGTERM handler");

	args = set_options(argc, argv);

	switch (command_opt) {
	case cmd_wenv:
		shell_worker_env();
		break;
	case cmd_init:
		if (args[1])
			init_repository(args[1]);
		else
			init_repository("alpine");
		break;
	}
	if (command_opt && args)
		return 0;

	/* locate bundled utilities */
	if ((renv_bin = findprog("renv")) == 0)
		not_found("renv");
	if ((rinstall_bin = findprog("rinstall")) == 0)
		not_found("rinstall");
	if ((rsub_bin = findprog("rsub")) == 0)
		not_found("rsub");

	/* subsequent operations are relative to the routes file */
	if (realpath(xdirname(routes_file), routes_realpath) == NULL)
		err(1, "realpath %s", routes_file);
	if (chdir(routes_realpath) == -1)
		err(1, "chdir %s", routes_realpath);

	/* try opening the routes file */
	if ((fd = open(routes_file, O_RDONLY)) == -1)
		err(1, "unable to open %s", routes_file);
	(void) close(fd);

	check_permissions(routes_realpath);

	if (!dryrun_opt) {
		/* require ssh-agent */
		if (verify_ssh_agent() != 0) {
			printf("Try running:\n");
			if (!getenv("SSH_AUTH_SOCK"))
				printf("  eval `ssh-agent`\n");
			printf("  ssh-add\n");
			exit(1);
		}

		/* Auto-upgrade utilities and verify path */
		install_if_new(renv_bin, REPLICATED_DIRECTORY "/renv");
		install_if_new(rinstall_bin, REPLICATED_DIRECTORY "/rinstall");
		install_if_new(rsub_bin, REPLICATED_DIRECTORY "/rsub");
		create_dir(ARCHIVE_DIRECTORY);
		create_dir(PUBLIC_DIRECTORY);
	}

	/* parse route labels */
	route_labels = alloc_labels();
	read_route_labels(routes_file);
	expand_route_labels();

	xregcomp(&label_reg, label_pattern, REG_EXTENDED);

	/* parse pln files for each host */
	for (n_routes = 0; route_labels[n_routes]; n_routes++)
		read_host_labels(route_labels[n_routes]);

	/* generate list of matching hostnames */
	hostnames = xcalloc(n_routes + 1, sizeof(char *), "hostnames");
	m_args = xcalloc(n_routes + 1, sizeof(char *), "m_args");
	compare_argv(args, hostnames, m_args);

	if (n_parallel > 0) {
		n_workers = 0;
		create_dir(log_directory);

		/* distribute hostnames */
		for (i = 0; hostnames[i]; i++) {
			j = i % n_parallel;
			if (j + 1 > n_workers) {
				worker_argv[n_workers] = xcalloc(argc, sizeof(char *), "worker_argv[]");
				worker_argc[n_workers] = create_worker_argv(argv, worker_argv[n_workers]);
				n_workers++;
			}
			worker_argv[j][worker_argc[j]++] = hostnames[i];
		}

		for (i = 0; i < n_workers; i++)
			worker_pid[i] = exec_worker(log_directory, i + 1, worker_argv[i]);

		rexec_summary(n_workers, worker_pid, log_directory);
		exit(0);
	}

	/* select a port to communicate on */
	http_port = get_socket();

	if (pledge("stdio rpath wpath cpath proc exec unveil", NULL) == -1)
		err(1, "pledge");

	/* main loop */
	if (dryrun_opt) {
		ret = dry_run(hostnames, m_args, &label_reg);
		free(hostnames);
		return ret;
	}

	ret = execute_remote(hostnames, &label_reg);
	free(hostnames);
	return ret;
}

/*
 * Execute commands on remote hosts
 * Returns an exit status
 */

static int
execute_remote(char *hostnames[], regex_t *label_reg) {
	char httpd_log[32768];
	int i, j, k, l;
	int nr;
	int ret;
	int stdout_pipe[2];
	size_t len;
	regmatch_t regmatch;
	Label **host_labels;

	int exit_code = 0;
	int scp_exit_code = 0;

	char *host_connect_msg = HL_HOST "%h" HL_RESET;
	char *host_connect_error_msg = HL_ERROR "%h initialization error" HL_RESET;
	char *label_exec_begin_msg = HL_LABEL "%l" HL_RESET;
	char *label_exec_end_msg = 0;
	char *label_exec_error_msg = HL_ERROR "%l exited with code %e" HL_RESET;
	char *host_disconnect_msg = 0;

	/* start background web server */
	start_http_server(stdout_pipe, http_port);

	/* custom log format */
	if (getenv("RSET_HOST_CONNECT")) {
		host_connect_msg = getenv("RSET_HOST_CONNECT");
		host_connect_error_msg = getenv("RSET_HOST_CONNECT_ERROR");
		label_exec_begin_msg = getenv("RSET_LABEL_EXEC_BEGIN");
		label_exec_end_msg = getenv("RSET_LABEL_EXEC_END");
		host_disconnect_msg = getenv("RSET_HOST_DISCONNECT");
		label_exec_error_msg = getenv("RSET_LABEL_EXEC_ERROR");
	}

	for (i = 0; route_labels[i]; i++) {
		host_labels = route_labels[i]->labels;

		for (k = 0; hostnames[k]; k++) {
			for (l = 0; l < route_labels[i]->n_aliases; l++) {
				if (strcmp(hostnames[k], route_labels[i]->aliases[l]) == 0)
					hostname = route_labels[i]->aliases[l];
				else
					continue;

				generate_session_id();
				log_msg(host_connect_msg, hostname, "", 0);

				len = PLN_LABEL_SIZE + sizeof(LOCAL_CONTROL_SOCKET);
				socket_path = xmalloc(len, "socket_path");
				snprintf(socket_path, len, LOCAL_CONTROL_SOCKET, hostname);

				ret = start_connection(
				    socket_path, hostname, route_labels[i], http_port, sshconfig_file);
				if (ret != 0) {
					log_msg(host_connect_error_msg, hostname, "", ret);
					end_connection(socket_path, hostname);
					free(socket_path);
					socket_path = NULL;
					continue;
				}

				for (j = 0; host_labels[j]; j++) {
					if (xregexec(label_reg, host_labels[j]->name, 1, &regmatch) != 0)
						continue;

					log_msg(label_exec_begin_msg, hostname, host_labels[j]->name, 0);

					/* local begin */
					local_exec(host_labels[j], host_labels[j]->options.begin);

					/* restore */
					if (restore_opt && host_labels[j]->export_paths[0])
						scp_exit_code = scp_archive(hostname, socket_path, host_labels[j], true);

					/* remote execution */
					if (tty_opt)
						exit_code =
						    ssh_command_tty(hostname, socket_path, host_labels[j], env_override);
					else
						exit_code =
						    ssh_command_pipe(hostname, socket_path, host_labels[j], env_override);

					/* archive */
					if (archive_opt && host_labels[j]->export_paths[0])
						scp_exit_code = scp_archive(hostname, socket_path, host_labels[j], false);

					/* local end */
					local_exec(host_labels[j], host_labels[j]->options.end);

					/* ssh terminated, unable to execute local interpreter */
					if ((exit_code == 255) || (exit_code == 127))
						log_msg(label_exec_error_msg, hostname, host_labels[j]->name, exit_code);
					else
						log_msg(label_exec_end_msg, hostname, host_labels[j]->name, exit_code);

					/* read output of web server */
					nr = read(stdout_pipe[0], httpd_log, sizeof(httpd_log));
					if (nr > 0) {
						httpd_log[nr] = '\0';
						trace_http(httpd_log);
					}
					if ((nr == -1) && (errno != EAGAIN))
						warn("read from httpd output");
				}

				if (socket_path) {
					if (archive_opt || restore_opt)
						log_msg(host_disconnect_msg, hostname, "", scp_exit_code);
					else
						log_msg(host_disconnect_msg, hostname, "", exit_code);
					end_connection(socket_path, hostname);
					free(socket_path);
					socket_path = NULL;
				}
			}
		}
	}

	return 0;
}

/*
 * Dry run: print hostnames and regex label matches
 * Returns an exit code
 */

static int
dry_run(char *hostnames[], char *m_args[], regex_t *label_reg) {
	int i, j, k, l;
	regmatch_t regmatch;
	regex_t route_reg;
	Label **host_labels;

	for (i = 0; route_labels[i]; i++) {
		host_labels = route_labels[i]->labels;

		for (k = 0; hostnames[k]; k++) {
			for (l = 0; l < route_labels[i]->n_aliases; l++) {
				if (strcmp(hostnames[k], route_labels[i]->aliases[l]) == 0)
					hostname = route_labels[i]->aliases[l];
				else
					continue;

				xregcomp(&route_reg, m_args[k], REG_EXTENDED);
				xregexec(&route_reg, hostname, 1, &regmatch);

				hl_range(hostname, HL_HOST, regmatch.rm_so, regmatch.rm_eo);
				printf("\n");

				for (j = 0; host_labels[j]; j++) {
					if (xregexec(label_reg, host_labels[j]->name, 1, &regmatch) != 0)
						continue;

					hl_range(host_labels[j]->name, HL_LABEL, regmatch.rm_so, regmatch.rm_eo);
					printf("\n");
				}
			}
		}
	}

	return 0;
}

/* signal handlers */

void
handle_exit(int sig) {
	if (socket_path && hostname && http_port) {
		printf("caught signal %d, terminating connection to '%s'\n", sig, hostname);
		/* clean up socket and SSH connection; leaving staging dir */
		execlp("ssh", "ssh", "-S", socket_path, "-O", "exit", hostname, NULL);
		err(1, "ssh -O exit");
	}
}

/* internal utilty functions */

static void
usage(bool summary) {
	fprintf(stderr, "release: %s\n", RELEASE);
	fprintf(stderr,
	    "usage: rset [-AnRt] [-E environment] [-F sshconfig_file] [-f routes_file]\n"
	    "            [-x label_pattern] hostname ...\n"
	    "       rset [-E environment] [-F sshconfig_file] [-f routes_file]\n"
	    "            [-x label_pattern] -o log_directory -p workers hostname ...\n"
	    "       rset -- command\n");
	if (!summary) {
		fprintf(stderr, "hint: use -h to display option summary\n");
		goto end;
	}

	printf("summary:\n"
	       "    -A                 Download files listed in label export paths\n"
	       "    -E environment     Key-value environment variables recognized by renv(1)\n"
	       "    -F sshconfig_file  Specify a ssh_config(5) file to use\n"
	       "    -f routes_file     Specify routes file using pln(5) format\n"
	       "    -n                 Print hostnames and matching labels\n"
	       "    -o log_directory   Log output of background tasks to directory\n"
	       "    -p workers         Run using parallel execution\n"
	       "    -R                 Upload files listed in label export paths\n"
	       "    -t                 Enable TTY input on remote host\n"
	       "    -x label_pattern   Execute labels matching specified regex\n");
	printf("commands:\n"
	       "    init [prefix]      Initialize a new repository matching example directory\n"
	       "    wenv               Print shell environment used by workers\n");
	printf("docs:\n"
	       "    man rset\n");

end:
	exit(1);
}

static char **
set_options(int argc, char *argv[]) {
	int ch;
	const char *errstr;
	opterr = 0;
	Options op;
	char argv_repr[10];

	bzero(&op, sizeof op);

	if (argv[1] && strcmp(argv[1], "-h") == 0)
		usage(true);

	/* commands */
	if (argv[1] && strcmp(argv[1], "--") == 0) {
		if (argc < 3)
			usage(false);

		if ((argc == 3 || argc == 4) && strcmp(argv[2], "init") == 0)
			command_opt = cmd_init;
		else if (argc == 3 && strcmp(argv[2], "wenv") == 0)
			command_opt = cmd_wenv;
		else {
			array_to_str(argv + 2, argv_repr, sizeof(argv_repr), " ");
			errx(1, "unknown command: '%s'", argv_repr);
		}

		return argv + 2;
	}

	while ((ch = getopt(argc, argv, "AnRtE:F:f:o:p:x:")) != -1) {
		switch (ch) {
		case 'A':
			archive_opt = 1;
			break;
		case 'n':
			dryrun_opt = 1;
			break;
		case 't':
			tty_opt = 1;
			break;
		case 'R':
			restore_opt = 1;
			break;
		case 'E':
			env_override = xstrdup(optarg, "env_override");
			env_split_lines(env_override);
			break;
		case 'F':
			sshconfig_file = optarg;
			break;
		case 'f':
			routes_file = optarg;
			break;
		case 'o':
			log_directory = optarg;
			break;
		case 'p':
			n_parallel = strtonum(optarg, 1, MAX_WORKERS, &errstr);
			if (errstr != NULL)
				errx(1, "number out of bounds %s: '%s'", errstr, argv[optind - 1]);
			break;
		case 'x':
			label_pattern = optarg;
			break;

		default:
			usage(false);
		}
	}
	if (optind >= argc)
		usage(false);

	if (n_parallel) {
		if (dryrun_opt || tty_opt || archive_opt || restore_opt)
			usage(false);
	}

	if ((log_directory == NULL) ^ (n_parallel == 0))
		usage(false);

	return argv + optind;
}

/* construct a list of hostnames matching routes */

static void
compare_argv(char *args[], char *hostnames[], char *m_args[]) {
	int i, j, k;
	int labels_matched;
	int n_hosts = 0;
	const char *match;

	for (i = 0; args[i]; i++) {
		labels_matched = 0;
		for (j = 0; route_labels[j]; j++) {
			match = pattern_match(args[i], route_labels[j]->aliases[0]);

			/* scan aliases for an exact match */
			if (!match) {
				for (k = 1; k < route_labels[j]->n_aliases; k++) {
					if (strcmp(args[i], route_labels[j]->aliases[k]) == 0)
						match = route_labels[j]->aliases[k];
				}
			}
			/* skip if already exists to set of hosts */
			for (k = 0; match && k < n_hosts; k++) {
				if (strcmp(match, hostnames[k]) == 0) {
					match = NULL;
					labels_matched++;
				}
			}
			/* add to list */
			if (match) {
				hostnames[n_hosts] = (char *) match;
				m_args[n_hosts] = args[i];
				n_hosts++;
				labels_matched++;
			}
		}
		if (labels_matched == 0)
			errx(1, "No match for '%s' in %s", args[i], routes_file);
	}

	hostnames[n_hosts] = NULL;
	m_args[n_hosts] = NULL;
}

/* failure to locate utility */

static void
not_found(char *name) {
	errx(1, "'%s' not found in PATH", name);
}

/* built-in http server */

static void
start_http_server(int stdout_pipe[], int http_port) {
	int flags;
	int status;
	char port[6];
	char *http_srv_argv[5];
	char *httpd_bin;
	pid_t http_server_pid;
	pid_t rset_pid;
	sigset_t set;

	snprintf(port, sizeof(port), "%u", http_port);
	http_srv_argv[0] = "miniquark";
	http_srv_argv[1] = "-d";
	http_srv_argv[2] = PUBLIC_DIRECTORY;
	http_srv_argv[3] = port;
	http_srv_argv[4] = NULL;

	if ((httpd_bin = findprog(http_srv_argv[0])) == 0)
		not_found(http_srv_argv[0]);

	/* start the web server */
	xpipe(stdout_pipe, "stdout");
	http_server_pid = fork();
	if (http_server_pid == 0) {
		/* close input side of pipe, and connect stdout */
		dup2(stdout_pipe[1], STDOUT_FILENO);
		close(stdout_pipe[1]);
		if (unveil(xdirname(PUBLIC_DIRECTORY), "r") == -1)
			err(1, "unveil");
		if (unveil(xdirname(httpd_bin), "x") == -1)
			err(1, "unveil");
		if (unveil("/usr/lib", "r") == -1)
			err(1, "unveil");
		if (unveil("/usr/libexec", "r") == -1)
			err(1, "unveil");
		if (pledge("stdio rpath proc exec", "stdio rpath proc inet") == -1)
			err(1, "pledge");

		execv(httpd_bin, http_srv_argv);
		fprintf(stderr, "Fatal: unable to start web server\n");
		err(1, "%s", httpd_bin);
	}

	/* close output side of pipe, and ensure readers don't block */
	close(stdout_pipe[1]);
	flags = fcntl(stdout_pipe[0], F_GETFL);
	fcntl(stdout_pipe[0], F_SETFL, flags | O_NONBLOCK);

	/* watchdog to ensure that the http server is shut down*/
	rset_pid = fork();
	if (rset_pid > 0) {
		if (pledge("stdio proc", NULL) == -1)
			err(1, "pledge");

		setproctitle("watch on pid %d", rset_pid);
		sigfillset(&set);
		sigprocmask(SIG_BLOCK, &set, NULL);
		if (waitpid(rset_pid, &status, 0) == -1)
			warn("wait on rset with pid %d", rset_pid);
		if (kill(http_server_pid, SIGTERM) == -1)
			err(1, "terminate http_server with pid %d", http_server_pid);
		exit(WEXITSTATUS(status));
	}
}

/*
 * Generate skeleton repository from share/examples/
 */
int
init_repository(const char *prefix) {
	char proj[DIR_MAX];
	int i;
	char *rset_bin, *bin_dir, *examples_dir;
	char *match;
	char search_path[3][DIR_MAX];
	char *cmd[5];

	if (getcwd(proj, DIR_MAX) == NULL)
		err(1, "getcwd");

	/* checks */
	assert_not_exists(".gitignore");
	assert_not_exists("routes.pln");

	/* infer examples path relative to rset binary */
	rset_bin = findprog("rset");
	if (rset_bin == NULL)
		errx(1, "rset not found in PATH");
	bin_dir = xdirname(rset_bin);

	/* possible locations */
	str_cpy(search_path[0], getenv("RSET_EXAMPLE_PATH"), DIR_MAX);
	snprintf(search_path[1], DIR_MAX, EXAMPLE_RELPATH, bin_dir);
	snprintf(search_path[2], DIR_MAX, EXAMPLE_RELPATH_ALT, bin_dir);

	for (i = 0; i < 3; i++) {
		if ((examples_dir = realpath(search_path[i], NULL)))
			break;
	}
	if (examples_dir == NULL)
		errx(1, "examples not found in search path:\n  %s\n  %s\n  %s",
		    search_path[0][0] ? search_path[0] : "RSET_EXAMPLE_PATH", search_path[1],
		    search_path[2]);

	match = list_dir_match(examples_dir, prefix);
	if (match == NULL)
		errx(1, "'%s' does not match any directory in %s", prefix, examples_dir);

	/* set permissions */
	if (chmod(proj, S_IRWXU) == -1)
		err(1, "chmod %s", proj);
	umask(026);

	/* standard directories */
	create_dir(REPLICATED_DIRECTORY);
	create_dir(ARCHIVE_DIRECTORY);
	create_dir(PUBLIC_DIRECTORY);

	/* copy */
	if (chdir(match) == -1)
		err(1, "chdir %s", match);
	cmd[0] = "cp";
	cmd[1] = "-r";
	cmd[2] = ".";
	cmd[3] = proj;
	cmd[4] = NULL;

	return run(cmd);
}
